mr/src/main/java/org/apache/iceberg/mr/mapred/IcebergInputFormat.java

```diff
@@ -0,0 +1,259 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.mr.mapred;
+
+import java.io.DataInput;
+import java.io.DataOutput;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Iterator;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.hive.ql.io.CombineHiveInputFormat;
+import org.apache.hadoop.mapred.FileSplit;
+import org.apache.hadoop.mapred.InputFormat;
+import org.apache.hadoop.mapred.InputSplit;
+import org.apache.hadoop.mapred.JobConf;
+import org.apache.hadoop.mapred.RecordReader;
+import org.apache.hadoop.mapred.Reporter;
+import org.apache.iceberg.CombinedScanTask;
+import org.apache.iceberg.DataFile;
+import org.apache.iceberg.FileScanTask;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.data.Record;
+import org.apache.iceberg.hadoop.HadoopInputFile;
+import org.apache.iceberg.hadoop.HadoopTables;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.mr.SerializationUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * CombineHiveInputFormat.AvoidSplitCombination is implemented to correctly delegate InputSplit
+ * creation to this class. See: https://stackoverflow.com/questions/29133275/
+ * custom-inputformat-getsplits-never-called-in-hive
+ */
+public class IcebergInputFormat<T> implements InputFormat<Void, T>, CombineHiveInputFormat.AvoidSplitCombination {
```

Comment:
Yeah, true. Do we think anything other than Hive is going to use this InputFormat? If not then we could just move the whole thing over to `iceberg-hive`. Otherwise we can leave it here and make the change you suggest.

Reply:
I doubt anyone would use the mapred format, so I think it makes sense to only add a `HiveInputFormat`.

I'm not sure we want to add it to `iceberg-hive`, unless we don't add many dependencies. Right now, that module is for interacting with the Hive metastore so it is used in Spark and other places. If we don't need additional dependencies then having a `HiveInputFormat` class makes sense. But if we need to make that module pull in more, then we should consider a separate module for Hive integration.

Reply:
Yea I think it would make sense to have a separate module. Any ideas on the name for that?

Reply:
`iceberg-hive-exec`?

Let's make sure we need it. I think this currently adds `compileOnly` dependencies, so we should be okay to put it in `iceberg-hive`.

Reply:
OK, I'm a bit confused now. Do we want to keep this here, move it to `iceberg-hive` or create a new subproject? Bear in mind that there will be other classes like the Hive SerDe, StorageHandler etc. built around this in future. We were intending to put all that in `iceberg-hive` but if you'd prefer to have some other subproject that contains them all that's fine too.

Reply:
I think a `HiveInputFormat` is a good idea, as we've been using a lot of Hive specific stuff to get this IF working with Hive (things like the interface @rdsr mentioned above and some properties in the JobConf set by Hive for things like column projection)

Reply:
Let's keep it where it is for now. We can move it if we need Hive later.